File: Pharmaceutical_Company_Example/Behavioral/C++/Visitor.cpp
#include <iostream>
#include <string>

// Forward declaration of Visitor
class Visitor;

// Element Interface for Regulatory Steps
class RegulatoryStep {
public:
  virtual void accept(Visitor& visitor) = 0;
  virtual ~RegulatoryStep() = default;
};

// Concrete Elements
class DocumentationSubmission : public RegulatoryStep {
public:
  void accept(Visitor& visitor) override;
};

class ClinicalTrialReview : public RegulatoryStep {
public:
  void accept(Visitor& visitor) override;
};

class SafetyEvaluation : public RegulatoryStep {
public:
  void accept(Visitor& visitor) override;
};

// Visitor Interface
class Visitor {
public:
  virtual void visit_documentation_submission(DocumentationSubmission& documentation) = 0;
  virtual void visit_clinical_trial_review(ClinicalTrialReview& trial_review) = 0;
  virtual void visit_safety_evaluation(SafetyEvaluation& safety_evaluation) = 0;
  virtual ~Visitor() = default;
};

// Concrete Visitor Auditor
class Auditor : public Visitor {
public:
  void visit_documentation_submission(DocumentationSubmission& documentation) override {
    std::cout << "Auditing documentation submission..." << std::endl;
  }
  
  void visit_clinical_trial_review(ClinicalTrialReview& trial_review) override {
    std::cout << "Auditing clinical trial review process..." << std::endl;
  }
  
  void visit_safety_evaluation(SafetyEvaluation& safety_evaluation) override {
    std::cout << "Auditing safety evaluation..." << std::endl;
  }
};

// Concrete Visitor Report Generator
class ReportGenerator : public Visitor {
public:
  void visit_documentation_submission(DocumentationSubmission& documentation) override {
    std::cout << "Generating report for documentation submission..." << std::endl;
  }
  
  void visit_clinical_trial_review(ClinicalTrialReview& trial_review) override {
    std::cout << "Generating report for clinical trial review..." << std::endl;
  }
  
  void visit_safety_evaluation(SafetyEvaluation& safety_evaluation) override {
    std::cout << "Generating report for safety evaluation..." << std::endl;
  }
};

// Concrete Visitor Approval Checker
class ApprovalChecker : public Visitor {
public:
  void visit_documentation_submission(DocumentationSubmission& documentation) override {
    std::cout << "Checking approval status of documentation submission..." << std::endl;
  }
  
  void visit_clinical_trial_review(ClinicalTrialReview& trial_review) override {
    std::cout << "Checking approval status of clinical trial review..." << std::endl;
  }
  
  void visit_safety_evaluation(SafetyEvaluation& safety_evaluation) override {
    std::cout << "Checking approval status of safety evaluation..." << std::endl;
  }
};

// Define accept functions for elements
void DocumentationSubmission::accept(Visitor& visitor) {
  visitor.visit_documentation_submission(*this);
}

void ClinicalTrialReview::accept(Visitor& visitor) {
  visitor.visit_clinical_trial_review(*this);
}

void SafetyEvaluation::accept(Visitor& visitor) {
  visitor.visit_safety_evaluation(*this);
}

int main() {
  // Define regulatory steps
  DocumentationSubmission documentation;
  ClinicalTrialReview clinical_trials;
  SafetyEvaluation safety_eval;
  
  // Define visitors
  Auditor auditor;
  ReportGenerator report_generator;
  ApprovalChecker approval_checker;
  
  // Use the Auditor visitor on different steps
  documentation.accept(auditor);
  clinical_trials.accept(auditor);
  safety_eval.accept(auditor);
  
  // Use the Report Generator visitor on different steps
  documentation.accept(report_generator);
  clinical_trials.accept(report_generator);
  safety_eval.accept(report_generator);
  
  // Use the Approval Checker visitor on different steps
  documentation.accept(approval_checker);
  clinical_trials.accept(approval_checker);
  safety_eval.accept(approval_checker);
  
  return 0;
}

File: Pharmaceutical_Company_Example/03_Behavioral/PHP/Visitor.php
<?php

// Define Element Interface for Regulatory Steps
interface RegulatoryStep
{
    public function accept(Visitor $visitor);
}

// Concrete Elements
class DocumentationSubmission implements RegulatoryStep
{
    public function accept(Visitor $visitor)
    {
        $visitor->visitDocumentationSubmission($this);
    }
}

class ClinicalTrialReview implements RegulatoryStep
{
    public function accept(Visitor $visitor)
    {
        $visitor->visitClinicalTrialReview($this);
    }
}

class SafetyEvaluation implements RegulatoryStep
{
    public function accept(Visitor $visitor)
    {
        $visitor->visitSafetyEvaluation($this);
    }
}

// Visitor Interface
interface Visitor
{
    public function visitDocumentationSubmission(DocumentationSubmission $documentation);
    public function visitClinicalTrialReview(ClinicalTrialReview $trialReview);
    public function visitSafetyEvaluation(SafetyEvaluation $safetyEvaluation);
}

// Concrete Visitor Auditor
class Auditor implements Visitor
{
    public function visitDocumentationSubmission(DocumentationSubmission $documentation)
    {
        echo "Auditing documentation submission...\n";
    }

    public function visitClinicalTrialReview(ClinicalTrialReview $trialReview)
    {
        echo "Auditing clinical trial review process...\n";
    }

    public function visitSafetyEvaluation(SafetyEvaluation $safetyEvaluation)
    {
        echo "Auditing safety evaluation...\n";
    }
}

// Concrete Visitor Report Generator
class ReportGenerator implements Visitor
{
    public function visitDocumentationSubmission(DocumentationSubmission $documentation)
    {
        echo "Generating report for documentation submission...\n";
    }

    public function visitClinicalTrialReview(ClinicalTrialReview $trialReview)
    {
        echo "Generating report for clinical trial review...\n";
    }

    public function visitSafetyEvaluation(SafetyEvaluation $safetyEvaluation)
    {
        echo "Generating report for safety evaluation...\n";
    }
}

// Concrete Visitor Approval Checker
class ApprovalChecker implements Visitor
{
    public function visitDocumentationSubmission(DocumentationSubmission $documentation)
    {
        echo "Checking approval status of documentation submission...\n";
    }

    public function visitClinicalTrialReview(ClinicalTrialReview $trialReview)
    {
        echo "Checking approval status of clinical trial review...\n";
    }

    public function visitSafetyEvaluation(SafetyEvaluation $safetyEvaluation)
    {
        echo "Checking approval status of safety evaluation...\n";
    }
}

// Define the regulatory steps
$documentation = new DocumentationSubmission();
$clinicalTrials = new ClinicalTrialReview();
$safetyEval = new SafetyEvaluation();

// Define visitors
$auditor = new Auditor();
$reportGenerator = new ReportGenerator();
$approvalChecker = new ApprovalChecker();

// Use the Auditor visitor on different steps
$documentation->accept($auditor);
$clinicalTrials->accept($auditor);
$safetyEval->accept($auditor);

// Use the Report Generator visitor on different steps
$documentation->accept($reportGenerator);
$clinicalTrials->accept($reportGenerator);
$safetyEval->accept($reportGenerator);

// Use the Approval Checker visitor on different steps
$documentation->accept($approvalChecker);
$clinicalTrials->accept($approvalChecker);
$safetyEval->accept($approvalChecker);

?>
